sensors/icinga2_state_change_events.py

```diff
@@ -1,26 +1,67 @@
+    urllib3.disable_warnings()  # pylint: disable=no-member
+except ImportError:
+    pass
 
 
 class Icinga2StateChangeSensor(Sensor):
     def setup(self):
         self.logger = self._sensor_service.get_logger(__name__)
-        self.api_url = self._config['api_url'] + '/events?queue=state_change&types=StateChange'
-        self.api_user = self._config['api_state_change_user']
-        self.api_password = self._config['api_state_change_password']
         self.trigger_name = 'event.state_change'
         self.trigger_pack = 'icinga2'
         self.trigger_ref = '.'.join([self.trigger_pack, self.trigger_name])
-        self.client = Client(self, self.api_url, self.api_user, self.api_password)
-        self.logger.info(
-            'Icinga2StateChangeSensor initialized with URL: %s User: %s Password: *****',
-            self.api_url, self.api_user)
+        self.types = ['StateChange']
+        self.queue = 'state_change'
+
+        self.api_url = self._config['api_url']
+
+        if ('api_user' in self._config and self._config['api_user'] is not None and  # noqa: W504
+                self._config['api_user']):
+            if 'api_password' in self._config and self._config['api_password'] is not None:
+                self.api_user = self._config['api_user']
+                self.api_password = self._config['api_password']
+                self.client = Client(self.api_url, self.api_user, self.api_password)
+                self.logger.info('Icinga2StateChangeSensor initialized with URL: %s User: %s',
+                                 self.api_url, self.api_user)
+            else:
+                raise ValueError("Username defined with no password.")
+        elif ('certificate' in self._config and self._config['certificate'] is not None
+              and self._config['certificate']):  # noqa: W503
+            self.certificate = self._config['certificate']
+            self.key = self._config.get('key', '')
+            self.ca_certificate = self._config.get('ca_certificate', '')
+            self.client = Client(self.api_url, certificate=self.certificate, key=self.key,
+                                 ca_certificate=self.ca_certificate)
+            self.logger.info('Icinga2StateChangeSensor initialized with URL: %s Certificate: %s',
+                             self.api_url, self.certificate)
+        else:
+            raise ValueError("Failed finding authentication method\n \
+                     Please specify either username and password or certificate location")
+
+    def run(self):
+        while True:
+            try:
+                self.logger.info('Connecting to event stream API.')
+                for event in self.client.events.subscribe(self.types, self.queue):
+                    self.process_event(event)
+            except Exception as ex:
+                self.logger.info('Icinga2StateChangeSensor Exception %s', str(ex))
+
+            # Wait between retries if Icinga API is disconnected
+            sleep(5)
```

Comment:
Should use a non blocking sleep here and ``eventlet.sleep``.

Reply:
Like 
https://github.com/StackStorm-Exchange/stackstorm-icinga2/pull/9/commits/3a7f540f04bd6f4d427a60e40d561c32b8b1f54f ?

Have managed to get away with not using `eventlet` before